### PR TITLE
Make sha1, s3_etag optional in TDR checksums (#2392)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [x] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

**Review by danielsotirhos — APPROVED** [review]
Quick overview looks good.
Fast-tracking ticket to @hannes-ucsc for review

File: src/azul/plugins/repository/tdr/__init__.py [hunk 1]
```diff
@@ -329,7 +330,10 @@ def without_values(cls) -> Dict[str, str]:
 
     @classmethod
     def extract(cls, json: JSON) -> 'Checksums':
-        return cls(**{f: json[f] for f in cls._fields})
+        return cls(**{
+            f: json.get(f) if is_optional(type_) else json[f]
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:334]: We should account not just for absence but also None values.

File: src/azul/types.py [hunk 2]
```diff
@@ -70,3 +71,24 @@ def get_remaining_time_in_millis(self) -> int:
...
+    :param type_: A type or type annotation.
+    :return: Whether the provided type is equivalent to typing.Optional[]
+    >>> is_optional(int)
+    False
+    >>> is_optional(type(None))
+    False
+    >>> from typing import Union
+    >>> is_optional(Union[int, float])
+    False
+    >>> from typing import Union
```
Comment (hannes-ucsc) [inline src/azul/types.py:87]: redundant

File: src/azul/types.py [hunk 3]
```diff
@@ -70,3 +71,24 @@ def get_remaining_time_in_millis(self) -> int:
 
     def log(self, msg: str) -> None:
         raise NotImplementedError
+
+
+def is_optional(type_) -> bool:
+    """
+    :param type_: A type or type annotation.
```
Comment (hannes-ucsc) [inline src/azul/types.py:78]: missing vertical space

File: src/azul/types.py (hunk 2 above)
Comment (hannes-ucsc) [inline src/azul/types.py:87]: redundant

File: src/azul/types.py [hunk 4]
```diff
@@ -70,3 +71,24 @@ def get_remaining_time_in_millis(self) -> int:
...
+    >>> is_optional(Union[int, float])
+    False
+    >>> from typing import Union
+    >>> is_optional(Union[int, None])
+    True
+    >>> from typing import Optional
+    >>> is_optional(Optional[int])
+    True
+    """
+    return typing.get_origin(type_) is Union and type(None) in typing.get_args(type_)
```
Comment (hannes-ucsc) [inline src/azul/types.py:94]: I like this better: https://stackoverflow.com/questions/56832881/check-if-a-field-is-typing-optional/62641842#62641842

File: src/azul/types.py [hunk 5]
```diff
@@ -70,3 +71,24 @@ def get_remaining_time_in_millis(self) -> int:
 
     def log(self, msg: str) -> None:
         raise NotImplementedError
+
+
+def is_optional(type_) -> bool:
```
Comment (hannes-ucsc) [inline src/azul/types.py:76]: I like this approach in general and I came up with sth very similar for the API, before even seeing this, I swear.

**Review by hannes-ucsc — COMMENTED** [review]
I'll fix these things

**Comment by hannes-ucsc** [comment]
Oh, the `add doctest` commit should have been a fixup.

File: src/azul/types.py [hunk 6]
```diff
@@ -1,3 +1,4 @@
+import typing
```
Comment (hannes-ucsc) [inline src/azul/types.py:1]: We shouldn't combine whole module imports and individual symbol imports from the same module.

File: src/azul/plugins/repository/tdr/__init__.py [hunk 7]
```diff
@@ -314,22 +318,42 @@ def retry_after(self) -> Optional[int]:
...
-        return self._asdict()
+    def to_json(self) -> Dict[str, str]:
+        """
+        >>> Checksums(crc32c='a', sha1='b', sha256='c', s3_etag=None).to_json()
+        {'crc32c': 'a', 'sha1': 'b', 'sha256': 'c'}
+        """
+        return {k: v for k, v in attr.asdict(self).items() if v is not None}
 
     @classmethod
-    def without_values(cls) -> Dict[str, str]:
```
Comment (hannes-ucsc) [inline src/azul/plugins/repository/tdr/__init__.py:327]: In your version of this PR, the tests still injected the optional fields with None values into the JSON representation of the descriptors, even though the real world descriptors will completely lack them. That's a case in point for #2374. It wasn't exactly fun to untangle the "forward" emulation in the test and consolidate that with the "backward" emulation in the CUT without knowing for sure that I didn't break anything.

**Comment by hannes-ucsc** [comment]
Looks like you checked `Ran make requirements_update or this PR leaves requirements*.txt, common.mk and Makefile untouched` but did neither, @noah-aviel-dove.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
Reindexing not finished yet, but far enough along, without fails.

**Pull request merged**